### ISSUE TITLE
why ExtractVolumePatches uses different implements for ThreadPoolDevice and GPUDevice.

### ISSUE THREAD

**Issue by quintinwang5** [tensorflow/tensorflow#46281]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Linux Ubuntu 20.04
- TensorFlow installed from (source or binary):source
- TensorFlow version (use command below):2.5
- Python version:3.6
- Bazel version (if compiling from source):3.7.1
- GCC/Compiler version (if compiling from source):7.5.0
- CUDA/cuDNN version:10.1
- GPU model and memory:7.0GB


**Describe the current behavior**
in tensorflow/core/kernels/eigen_volume_patch.h
```c++
650 OVERRIDE_EVALUATOR(Eigen::ThreadPoolDevice);
651 OVERRIDE_EVALUATOR(Eigen::DefaultDevice);
```
Overrides the TensorExecutor to modify the PADDING_SAME to
```c++
  131         case PADDING_SAME: {                                                                                                                                                                                                                     
  132           m_outputPlanes = Eigen::divup(m_input_planes_eff, m_plane_strides);
  133           m_outputRows = Eigen::divup(m_input_rows_eff, m_row_strides);
  134           m_outputCols = Eigen::divup(m_input_cols_eff, m_col_strides);
  135           const Index dz = numext::maxi<DenseIndex>(
  136               0, (m_outputPlanes - 1) * m_plane_strides + m_patch_planes_eff -
  137                      m_input_planes_eff);
  138           const Index dy = numext::maxi<DenseIndex>(
  139               0, (m_outputRows - 1) * m_row_strides + m_patch_rows_eff -
  140                      m_input_rows_eff);
  141           const Index dx = numext::maxi<DenseIndex>(
  142               0, (m_outputCols - 1) * m_col_strides + m_patch_cols_eff -
  143                      m_input_cols_eff);
  144           m_planePaddingTop = dz / 2;
  145           m_rowPaddingTop = dy / 2;
  146           m_colPaddingLeft = dx / 2;
  147           break;
  148         }
```

but for GPUDevice, we don't do this override. We use default eigen implement:
```c++
  257         case PADDING_SAME: {                                                                                                                                                                                                                     
  258           m_outputPlanes = numext::ceil(m_input_planes_eff / static_cast<float>(m_plane_strides));
  259           m_outputRows = numext::ceil(m_input_rows_eff / static_cast<float>(m_row_strides));
  260           m_outputCols = numext::ceil(m_input_cols_eff / static_cast<float>(m_col_strides));
  261           const Index dz = m_outputPlanes * m_plane_strides + m_patch_planes_eff - 1 - m_input_planes_eff;
  262           const Index dy = m_outputRows * m_row_strides + m_patch_rows_eff - 1 - m_input_rows_eff;
  263           const Index dx = m_outputCols * m_col_strides + m_patch_cols_eff - 1 - m_input_cols_eff;
  264           m_planePaddingTop = dz - dz / 2;
  265           m_rowPaddingTop = dy - dy / 2;
  266           m_colPaddingLeft = dx - dx / 2;
  267           break;
  268         }
```

The question is: Why do we use different implements for CPU/GPU devices?

**Describe the expected behavior**
Different devices should produce same results.

**Comment by sanjoy** [comment]
@quintinwang5, just so I understand are you saying that TF produces substantially different results on CPU vs GPU?  If yes, can you demonstrate this using a Python snippet?

**Comment by quintinwang5** [comment]
@sanjoy The code is here. If we use `padding="SAME"`. The result will be different. And `padding="VALID"` works well.
``` python
#!/usr/bin/env python
# coding=utf-8
import tensorflow as tf
import numpy as np
tf.compat.v1.disable_eager_execution() # need to disable eager in TF2.x
with tf.device("/CPU:0"):
    a = tf.random.normal(shape=[6, 9, 7, 8, 4], seed=1)
    b = tf.random.normal(shape=[6, 9, 7, 8, 4], seed=1)
    c = tf.extract_volume_patches(a, ksizes=[1, 1, 1, 1, 1], strides=[1, 2, 3, 4, 1], padding="SAME")
with tf.device("/GPU:0"):
    d = tf.extract_volume_patches(b, ksizes=[1, 1, 1, 1, 1], strides=[1, 2, 3, 4, 1], padding="SAME")


sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(allow_soft_placement=False, log_device_placement=True))

print(sess.run(tf.reduce_all(tf.equal(c, d))))

```

**Comment by sanjoy** [comment]
I don't have enough context here, but it looks like @mjanusz added the override.  Michal any ideas?  Does the GPU kernel need some changes?

**Comment by tilakrayal** [comment]
@quintinwang5,
I tried to execute the mentioned code on tensorflow v2.15 and it was executed without any issue/error and also observed that with both `padding='SAME'` and `'VALID'` the output is also similar. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/c2221f51ed5d1defb74b6eb7f62e6139/untitled1676.ipynb). Thank you!
